src/app/clusters/thermostat-server/thermostat-server.cpp

```diff
@@ -1287,102 +1322,126 @@ bool emberAfThermostatClusterSetActivePresetRequestCallback(
+            {
+                return false;
+            }
+            requestedPresets = true;
+            break;
+        case Schedules::Id:
+            if (requestedSchedules) // Double-requesting an attribute is invalid
+            {
+                return false;
+            }
+            requestedSchedules = true;
+            break;
+        default:
+            return false;
+        }
+    }
+    if (attributeIdsIter.GetStatus() != CHIP_NO_ERROR)
     {
-        ExtendTimer(endpoint, commandData.timeoutSeconds);
-        commandObj->AddStatus(commandPath, imcode::Success);
-        return true;
+        return false;
     }
-
-    // Set presets editable to true and the scoped originator node id to the source scoped node id, and start a timer with the
-    // timeout in seconds passed in the command args. Return success.
-    gThermostatAttrAccess.SetPresetsEditable(endpoint, true);
-    gThermostatAttrAccess.SetOriginatorScopedNodeId(endpoint, sourceNodeId);
-    ScheduleTimer(endpoint, commandData.timeoutSeconds);
-    commandObj->AddStatus(commandPath, imcode::Success);
-    return true;
+    // If the atomic request doesn't contain at least one of these attributes, it's invalid
+    return (requestedPresets || requestedSchedules);
 }
 
-bool emberAfThermostatClusterCancelPresetsSchedulesEditRequestCallback(
-    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
-    const chip::app::Clusters::Thermostat::Commands::CancelPresetsSchedulesEditRequest::DecodableType & commandData)
+bool handleAtomicBegin(chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+                       const Commands::AtomicRequest::DecodableType & commandData)
 {
     EndpointId endpoint = commandPath.mEndpointId;
 
-    // If presets are not editable, return INVALID_IN_STATE.
-    if (!gThermostatAttrAccess.GetPresetsEditable(endpoint))
+    if (gThermostatAttrAccess.InAtomicWrite(commandObj, endpoint))
     {
+        // This client already has an open atomic write
         commandObj->AddStatus(commandPath, imcode::InvalidInState);
         return true;
     }
 
-    ScopedNodeId sourceNodeId = GetSourceScopedNodeId(commandObj);
-
-    // If the node id sending the CancelPresetsSchedulesRequest command is not the same as the one which send the
-    // previous StartPresetsSchedulesEditRequest, return UNSUPPORTED_ACCESS.
-    if (gThermostatAttrAccess.GetOriginatorScopedNodeId(endpoint) != sourceNodeId)
+    if (!commandData.timeout.HasValue())
     {
-        commandObj->AddStatus(commandPath, imcode::UnsupportedAccess);
+        commandObj->AddStatus(commandPath, imcode::InvalidCommand);
         return true;
     }
 
-    Delegate * delegate = GetDelegate(endpoint);
+    auto timeout = commandData.timeout.Value();
 
-    if (delegate == nullptr)
+    if (!validAtomicAttributes(commandData))
```

Comment:
So per spec, just the way things are written, I would expect this check to come before the checks for an atomic write state and so on....  Followup to sort out the order here.